error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/MonoOfPublishers.java

```diff
@@ -0,0 +1,58 @@
+package tech.picnic.errorprone.bugpatterns;
+
+import static com.google.errorprone.BugPattern.LinkType.CUSTOM;
+import static com.google.errorprone.BugPattern.SeverityLevel.WARNING;
+import static com.google.errorprone.BugPattern.StandardTags.FRAGILE_CODE;
+import static tech.picnic.errorprone.bugpatterns.util.Documentation.BUG_PATTERNS_BASE_URL;
+import static tech.picnic.errorprone.bugpatterns.util.MoreTypes.generic;
+import static tech.picnic.errorprone.bugpatterns.util.MoreTypes.subOf;
+import static tech.picnic.errorprone.bugpatterns.util.MoreTypes.type;
+import static tech.picnic.errorprone.bugpatterns.util.MoreTypes.unbound;
+
+import com.google.auto.service.AutoService;
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.bugpatterns.BugChecker.MethodInvocationTreeMatcher;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.suppliers.Supplier;
+import com.google.errorprone.suppliers.Suppliers;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.tools.javac.code.Type;
+import org.reactivestreams.Publisher;
+import reactor.core.publisher.Mono;
+
+/**
+ * A {@link BugChecker} that flags nesting of {@link Publisher Publishers} inside {@link Mono
+ * Monos}.
+ */
+@AutoService(BugChecker.class)
+@BugPattern(
+    summary =
+        "Avoid nesting `Publisher`s inside `Mono`s; the resultant code is hard to reason about",
+    link = BUG_PATTERNS_BASE_URL + "MonoOfPublishers",
+    linkType = CUSTOM,
+    severity = WARNING,
+    tags = FRAGILE_CODE)
+public final class MonoOfPublishers extends BugChecker implements MethodInvocationTreeMatcher {
```

Comment:
is it possible to make this something like `NestedPublishers` instead? do we care if it's a `Mono` specifically?

Reply:
At this point, as pointed out by @Stephan202 in the original comment (That I can't link here as it is a private repo), we do not want to flag Flux of publishers because this will warn about all `groupBy` operator usage.

Reply:
Indeed, it would effectively disallow [groupBy](https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Flux.html#groupBy-java.util.function.Function-) usage. To be fair, that is a very sharp tool indeed, but we do use it in places. In fact, the `FluxFlatMapUsage` has [special logic](https://github.com/PicnicSupermarket/error-prone-support/blob/c53a3f64b6ecf40e8b70f56e6bf5d4cb6d8f5acc/error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/FluxFlatMapUsage.java#L90-L99) for `Flux`es of nested publishes.

Reply:
Ah yeah the `GroupedFlux` 🤪 Wouldn't it be possible to make an exception for it here as well though?

Reply:
@mohamedsamehsalah it's the same in the example shared by @Stephan202 right? again I'm not 100% sure it's possible, but [there](https://github.com/PicnicSupermarket/error-prone-support/blob/c53a3f64b6ecf40e8b70f56e6bf5d4cb6d8f5acc/error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/FluxFlatMapUsage.java#L90-L99) we first check if `FLUX_FLATMAP ` matches and then do a more specific check (which there goes to 2 different cases, but here 1 case could go to `Description.NO_MATCH` right?)

Reply:
> Is this even feasible, since we are matching on the Class and generic level 

We can specify anything we want by using the methods from `MoreTypes`. There is a nice test that allows us to more clearly see how we can use these types to match specific types with generics or [unbounded](https://www.tutorialspoint.com/what-are-unbounded-wildcard-w-r-t-generics-method-in-java#:~:text=An%20unbounded%20wildcard%20is%20the,to%20declare%20an%20unbounded%20wildcard).  

[Here](https://github.com/PicnicSupermarket/error-prone-support/blob/master/error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/util/MoreTypesTest.java#L122-L123) we have a test `BugChecker` that loops over the list of [test types](https://github.com/PicnicSupermarket/error-prone-support/blob/master/error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/util/MoreTypesTest.java#L154) and goes over all the `MethodInvocation`s [here](https://github.com/PicnicSupermarket/error-prone-support/blob/master/error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/util/MoreTypesTest.java#L44-L111) and reports method invocations that are a subtype of any type defined in the list of test types.

Okay this get's rather complicated and sounds more daunting than it is to check for `GroupedFlux` in an if statement, that should be doable. I think @Venorcis  is right and we can do something like that. That would make the check more generic as well as we would also consider `Flux`es. 

If you are like: hmm not sure if I fully understand & this gets a little more complicated than expected, I can just swing by and we can look at it together 😄. That is totally fine.

Reply:
> If you are like: hmm not sure if I fully understand & this gets a little more complicated than expected...

I do need time to digest this indeed 😅 ..

Reply:
See [a27838d](https://github.com/PicnicSupermarket/error-prone-support/pull/642/commits/a27838d952e7fe86e64c864ae58545b9ad484b5e).